src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,520 @@
+#include "join_mpsm.hpp"
+
+#include <algorithm>
+#include <map>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "join_mpsm/radix_cluster_sort_numa.hpp"
+#include "resolve_type.hpp"
+#include "scheduler/abstract_task.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "scheduler/job_task.hpp"
+#include "storage/column_visitable.hpp"
+#include "storage/dictionary_column.hpp"
+#include "storage/numa_placement_manager.hpp"
+#include "storage/reference_column.hpp"
+#include "storage/value_column.hpp"
+
+STRONG_TYPEDEF(size_t, ClusterID);
+
+namespace opossum {
+JoinMPSM::JoinMPSM(const std::shared_ptr<const AbstractOperator> left,
+                   const std::shared_ptr<const AbstractOperator> right, const JoinMode mode,
+                   const std::pair<ColumnID, ColumnID>& column_ids, const PredicateCondition op)
+    : AbstractJoinOperator(OperatorType::JoinMPSM, left, right, mode, column_ids, op) {
+  // Validate the parameters
+  DebugAssert(mode != JoinMode::Cross, "This operator does not support cross joins.");
+  DebugAssert(left != nullptr, "The left input operator is null.");
+  DebugAssert(right != nullptr, "The right input operator is null.");
+  DebugAssert(op == PredicateCondition::Equals, "Only Equi joins are supported by MPSM join.");
+}
+
+std::shared_ptr<const Table> JoinMPSM::_on_execute() {
+  // Check column types
+  const auto& left_column_type = input_table_left()->column_data_type(_column_ids.first);
+  DebugAssert(left_column_type == input_table_right()->column_data_type(_column_ids.second),
+              "Left and right column types do not match. The mpsm join requires matching column types");
+
+  // Create implementation to compute the join result
+  _impl = make_unique_by_data_type<AbstractJoinOperatorImpl, JoinMPSMImpl>(
+      left_column_type, *this, _column_ids.first, _column_ids.second, _predicate_condition, _mode);
+
+  return _impl->_on_execute();
+}
+
+void JoinMPSM::_on_cleanup() { _impl.reset(); }
+
+std::shared_ptr<AbstractOperator> JoinMPSM::_on_recreate(
+    const std::vector<AllParameterVariant>& args, const std::shared_ptr<AbstractOperator>& recreated_input_left,
+    const std::shared_ptr<AbstractOperator>& recreated_input_right) const {
+  return std::make_shared<JoinMPSM>(recreated_input_left, recreated_input_right, _mode, _column_ids,
+                                    _predicate_condition);
+}
+
+const std::string JoinMPSM::name() const { return "Join MPSM"; }
+
+template <typename T>
+class JoinMPSM::JoinMPSMImpl : public AbstractJoinOperatorImpl {
+ public:
+  JoinMPSMImpl<T>(JoinMPSM& sort_merge_join, ColumnID left_column_id, ColumnID right_column_id,
+                  const PredicateCondition op, JoinMode mode)
+      : _mpsm_join{sort_merge_join},
```

Comment:
This might be confusing. What is the difference between “MPSM” and “sort_merge_join”?